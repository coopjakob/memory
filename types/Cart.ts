interface Price {
  currencyIso: String
  formattedValue: String
  priceType: String
  value: Number
}

interface CartProduct {
  code: String
  depositPrice: Price
  keywords: Array<any>
  name: String
  pickPrice: Price
  unit: {
    name: String
  }
  url: String
  variances: Array<any>
}

export interface CartState {
  type: String
  appliedOrderPromotions: Array<any>
  appliedProductPromotions: Array<any>
  code: String
  entries: Array<{
    entryNumber: Number
    product: CartProduct
    quantity: Number
    replace: Boolean
    totalPrice: Price
  }>
  extraAmountToBeReserved: Price
  guid: String
  isPartOfSubscription: Boolean
  postCode: String
  timeWindow: any
  totalDepositSum: Price
  totalItems: Number
  totalPrice: Price
  totalPriceWithExtraAmountToReserve: Price
  totalPriceWithTax: Price
  totalQuantity: Number
  totalTaxValues: Array<{
    appliedValue: Price
    value: Number
  }>
  amountMissing: Price
  coopStore: {
    code: String
    enova: Boolean
    foodBagsOnly: Boolean
    name: String
  }
  isExtraOrderCart: Boolean
  recipesGroups: Array<any>
  replaceAll: Boolean
}
