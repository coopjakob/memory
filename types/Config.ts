export interface ConfigState {
  cartguid?: String
  authToken?: String
  rrSessionId?: String
  coopStore?: String
  storeName?: String
  user?: String
  numberOfOrders?: Number
  hasSavedCarts?: Boolean
  medmeraId?: String
  b2bUser?: Boolean
  sessionPostCode?: String
  deliveryMode?: {
    code: String
    name: String
  }
}
