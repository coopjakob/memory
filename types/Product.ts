interface Price {
  formattedValue: string
}

interface Image {
  altText: string
  url: string
}

interface Promotion {
  maxUseText: boolean
  medmera: boolean
  description: string
}

interface Category {
  code: string
  name: string
}

export default interface Product {
  name: string
  images: Array<Image>
  url: string
  code: string
  manufacturer: string
  packageSizeInformation: string
  comparisonPrice: Price
  deposit: Price
  depositPrice: Price
  price: Price
  consumerInformationText: string
  fromSweden: boolean
  promotionPrice: Price
  potentialPromotions: Array<Promotion>
  categories: Array<Category>
  productLabels: Array<Category>
}
