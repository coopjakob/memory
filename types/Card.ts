import Product from '~/types/Product'

export enum CardTypes {
  INFO = 'info',
  PRODUCT = 'product',
  AD = 'ad'
}

export interface ProductCard extends Partial<Product> {
  type: CardTypes.PRODUCT
}

export type Cards = Array<ProductCard | ExtraCard>

export type ExtraCard = AdCard | InfoCard

interface GenericExtraCard {
  position?: number
  label?: string
  brand?: string
}

export interface AdCard extends GenericExtraCard {
  type: CardTypes.AD
  image: string
  link: string
}

export interface InfoCard extends GenericExtraCard {
  type: CardTypes.INFO
  image: string
  imageAltText: string
  header: string
  text: string
  link: string
  buttonText: string
}
