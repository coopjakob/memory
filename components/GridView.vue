<template>
  <div>
    <div ref="grid" class="grid">
      <template v-for="card in cards">
        <div
          v-if="didShowMore && Array.isArray(card)"
          :key="card.sortKey"
          class="buddy"
        >
          <component
            :is="components[_card.type]"
            v-for="_card in card"
            :key="_card.sortKey"
            ref="card"
            :card="_card"
          />
        </div>
        <component
          :is="components[card.type]"
          v-else
          :key="card.sortKey"
          ref="card"
          :card="card"
          :columns="columns"
        />
      </template>
      <component
        :is="components[card.type]"
        v-for="card in selectedFillers"
        :key="card.sortKey"
        :card="card"
        :columns="columns"
      />
    </div>

    <div v-if="!loading" class="show-more">
      <Coopknapp
        v-if="!didShowMore && isMobile"
        cta="Visa mer"
        :disabled="loading"
        @click="loadFull"
      />
    </div>
    <div v-else-if="isMobile" class="show-more loading">
      <div class="lds-ellipsis">
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import ProductCard from './ProductCard.vue'
import InfoCard from './InfoCard.vue'
import AdCard from './AdCard.vue'
import SkeletonCard from './SkeletonCard.vue'
import Coopknapp from './Coopknapp.vue'
import { CardTypes } from '~/types/Card'

export default Vue.extend({
  components: {
    Coopknapp
  },
  props: {
    columns: {
      type: Number,
      required: true
    },
    loading: Boolean,
    didShowMore: Boolean,
    isMobile: Boolean,
    width: {
      type: Number,
      required: true
    },
    cards: {
      type: Array,
      required: true
    },
    selectedFillers: {
      type: Array,
      required: true
    },
    emptySlots: {
      type: Number,
      required: true
    },
    loadFull: {
      type: Function,
      required: true
    }
  },
  data() {
    return {
      components: {
        [CardTypes.PRODUCT]: ProductCard,
        [CardTypes.SKELETON]: SkeletonCard,
        [CardTypes.INFO]: InfoCard,
        [CardTypes.AD]: AdCard
      }
    }
  }
})
</script>

<style>
/* creds to https://loading.io/css/ */

.lds-ellipsis {
  display: inline-block;
  position: relative;
  width: 80px;
  height: 80px;
}
.lds-ellipsis div {
  position: absolute;
  top: 33px;
  width: 13px;
  height: 13px;
  border-radius: 50%;
  background: #00aa46;
  animation-timing-function: cubic-bezier(0, 1, 1, 0);
}
.lds-ellipsis div:nth-child(1) {
  left: 8px;
  animation: lds-ellipsis1 0.6s infinite;
}
.lds-ellipsis div:nth-child(2) {
  left: 8px;
  animation: lds-ellipsis2 0.6s infinite;
}
.lds-ellipsis div:nth-child(3) {
  left: 32px;
  animation: lds-ellipsis2 0.6s infinite;
}
.lds-ellipsis div:nth-child(4) {
  left: 56px;
  animation: lds-ellipsis3 0.6s infinite;
}
@keyframes lds-ellipsis1 {
  0% {
    transform: scale(0);
  }
  100% {
    transform: scale(1);
  }
}
@keyframes lds-ellipsis3 {
  0% {
    transform: scale(1);
  }
  100% {
    transform: scale(0);
  }
}
@keyframes lds-ellipsis2 {
  0% {
    transform: translate(0, 0);
  }
  100% {
    transform: translate(24px, 0);
  }
}
</style>

<style lang="sass" scoped>
.grid
  display: grid
  grid-template-columns: repeat(auto-fill, minmax(150px, 1fr))
  grid-auto-rows: minmax(350px, auto)
  grid-auto-flow: dense
  grid-gap: 2px

.card
  box-sizing: border-box
  display: flex
  flex-direction: column
  position: relative

.buddy
  grid-column-end: span 2
  display: grid
  grid-template-columns: repeat(auto-fill, minmax(150px, 1fr))
  grid-gap: 2px

.show-more
  width: 100%
  text-align: center
</style>
