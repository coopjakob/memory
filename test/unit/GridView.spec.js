import { mount } from '@vue/test-utils'
import Vue from 'vue'
import GridView from '@/components/GridView.vue'
import { CardTypes } from '@/types/Card'

const generateSkeletons = (n) =>
  Array(n).fill({
    type: CardTypes.SKELETON
  })

const mountWithProps = (component, props = {}) => {
  return mount(component, {
    propsData: {
      columns: 3,
      loading: false,
      didShowMore: false,
      isMobile: false,
      width: 800,
      cards: generateSkeletons(10),
      selectedFillers: [],
      emptySlots: 5,
      loadFull: jest.fn(),
      ...props
    }
  })
}

describe('Grid.spec.js', () => {
  test('snapshots', () => {
    // On desktop
    const wrapper = mountWithProps(GridView)
    expect(wrapper.element).toMatchSnapshot()

    // On mobile
    const wrapper2 = mountWithProps(GridView, {
      isMobile: true
    })
    expect(wrapper2.element).toMatchSnapshot()

    // On mobile with more content shown
    const wrapper3 = mountWithProps(GridView, {
      isMobile: true,
      didShowMore: true
    })
    expect(wrapper3.element).toMatchSnapshot()
  })

  describe('show more desktop', () => {
    test('no show more on desktop', () => {
      const wrapper = mountWithProps(GridView)
      expect(
        wrapper
          .find('.show-more')
          .find('button')
          .exists()
      ).toBe(false)
    })
  })

  describe('show more mobile', () => {
    let wrapper // eslint-disable-line
    const func = jest.fn(() => {
      wrapper.setProps({
        loading: true
      })
    })
    wrapper = mountWithProps(GridView, {
      loadFull: func,
      isMobile: true,
      cards: generateSkeletons(6)
    })

    test('is there a button to show more', () => {
      expect(wrapper.vm.cards.length).toBe(6)
      expect(wrapper.find('.show-more')).toBeTruthy()
    })

    test('show loading after clicking', async () => {
      wrapper
        .find('.show-more')
        .find('button')
        .trigger('click')
      await Vue.nextTick()
      expect(func.mock.calls.length).toBe(1)
      expect(wrapper.vm.loading).toBe(true)
      expect(wrapper.vm.isMobile).toBe(true)
      expect(wrapper.find('.show-more.loading').exists()).toBe(true)
    })

    test('after load', async () => {
      wrapper.setProps({
        didShowMore: true,
        loading: false,
        cards: generateSkeletons(30)
      })
      await Vue.nextTick()
      expect(wrapper.find('.show-more.loading').exists()).toBe(false)
      expect(wrapper.vm.cards.length).toBeGreaterThan(6)
    })
  })
})
