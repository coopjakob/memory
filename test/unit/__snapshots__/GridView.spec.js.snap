// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid.spec.js snapshots 1`] = `
<div>
  <div
    class="grid"
  >
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
     
  </div>
   
  <div
    class="show-more"
  >
    <!---->
  </div>
</div>
`;

exports[`Grid.spec.js snapshots 2`] = `
<div>
  <div
    class="grid"
  >
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
     
  </div>
   
  <div
    class="show-more"
  >
    <button
      class="coopknapp"
    >
      
  Visa mer

    </button>
  </div>
</div>
`;

exports[`Grid.spec.js snapshots 3`] = `
<div>
  <div
    class="grid"
  >
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
    <div
      card="[object Object]"
      class="card"
      columns="3"
    >
      <div
        class="top"
      >
        <div
          class="image"
        />
         
        <div
          class="line"
        />
         
        <div
          class="line"
        />
      </div>
       
      <div
        class="bottom"
      >
        <div
          class="short-line"
        />
         
        <div
          class="button"
        />
      </div>
    </div>
     
  </div>
   
  <div
    class="show-more"
  >
    <!---->
  </div>
</div>
`;
